Bugfix 343: Replace XML-based preview cards with Composable overlay

Armetiis opened the pull request:
This PR addresses an intermittent issue where preview cards, previously implemented as XML resources, were not consistently displayed when the markers were clicked. This solution replaces the XML-based approach with a Composable overlay, improving reliability and maintainability.

### Changes

1. **Removed ViewAnnotationManager in the MapViewModel:**
   - Eliminated `ViewAnnotationManager` and related code.
   - Removed calls to `mapViewModel.updateViewAnnotationManager()` and `mapViewModel.removePreviewCard()`.

2. **Introduced Composable PreviewCard:**
   - Created a new `PreviewCard` Composable function to display parking information.
   - The card is positioned using a custom layout modifier
   - Introduced `showPreviewCard` state to control the visibility of the preview card.
   - Updated click listeners to set `showPreviewCard = true` when a parking is selected.

3. **Refactored Click Handlers:**
   - Created a common `handleAnnotationClick` function for both point and polygon annotations.
   - This function updates the selected parking, shows the preview card, and centers the map.
   
### Benefits

- **Improved Reliability:** Eliminates issues related to XML-based view annotations not displaying correctly.
- **Enhanced Information:** Users now see more details about each parking spot in the preview.
- **Modification potential:** Having a composable instead of a string resource makes it much easier to customize the card

### New Preview Card's UI
<img width="250" alt="image" src="https://github.com/user-attachments/assets/f6c41149-118a-422d-8c05-bfd5b4ad69d8">

### Testing and coverage
As you all know, it is not possible to interact with the map in the tests to select a parking. This makes it impossible to test the interaction with the markers and ensure the correct dispaly of the card. However, given that the card is now a composable, it was possible to test it in isolation by setting it as content of the compose test rule.
![image](https://github.com/user-attachments/assets/9626f116-dc72-4889-b455-468d9fcf4176)

-- -
- Closes #343

Pijukai commented:
First i want to say that the preview card looks great now so good job. Second we finally passed the 80% in map so congrats as well.

Pijukai commented on app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:
```diff
@@ -329,89 +325,46 @@ fun MapScreen(
-                        center(parkingDeserialized.location.center)
+                // ======================= ANNOTATION CLICK LISTENER =======================
+                // Define a common click handler function
+                fun handleAnnotationClick(
```
since it is a function maybe you can add the description for the parameters ?

Armetiis replied:
Sure, will add documentation on this

Pijukai commented on app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:
```diff
@@ -329,89 +325,46 @@ fun MapScreen(
+                // ======================= ANNOTATION CLICK LISTENER =======================
+                // Define a common click handler function
+                fun handleAnnotationClick(
+                    annotation: Any,
```
Here instead of using `Any` you could have used `Annotation` since it is the superclass of both PointAnnotation and PolygonAnnotation. This way we are more in control of what goes inside the function.

Armetiis replied:
I thought about that, but somehow it doesn't work. I get the following error: Incompatible types: PointAnnotation and Annotation. I don't really know how to do it

Pijukai replied:
The code is good as it is and the switch will handle this perfectly so it doesn't matter for this PR. Since you looked into i will not block the PR for it.

Pijukai commented on app/src/androidTest/java/com/github/se/cyrcle/ui/map/MapScreenTest.kt:
```diff
@@ -285,4 +286,25 @@ class MapScreenTest {
   }
+
+  @Test
+  fun testPreviewCard() {
```
You did the best that you could to test it so well done :+1:

Pijukai replied:
You could improve the test by asserting what is displayed on the card since it is linked to parking1 if I am not wrong.

Armetiis replied:
About this. The content of the preview card is clearly not the definitive one. I just did this to have an idea of what the Preview Card will look like. The point of this PR is not to design a Preview Card, but rather to ensure their reliability. This is something that can be addressed when we work on its content during the polish sprint.

Pijukai replied:
It was just a suggestion i agree with you and i will make sure to bring it up during the polishing.

Pijukai commented on app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:
```diff
@@ -163,6 +164,10 @@ fun MapScreen(
   // Mutable state to store the PointAnnotationManager for parking labels
   var pLabelAnnotationManager by remember { mutableStateOf<PointAnnotationManager?>(null) }
 
+  var showPreviewCard by remember { mutableStateOf(false) }
```
It is a non blocking issue but since all other variables have documentation maybe you can add one here ?

Armetiis replied:
Sure, i'll do it for consistency. But I believe that sometimes, trivial comments should be avoided. The name of the variable is self explanatory.

sean-prz reviewed (commented):
Thanks for this ! Another PR down the drain fixing problems with the map !
This will make adding the icons on the card next sprint so much easier. 

The code is clean, I will not waste our time on futilities. 
Instead take the time to fix and relevant problems : 

When we zoom with double click the card is not removed, and so we can move out and the card floats around without being on top of the parking : 

<img src ="https://github.com/user-attachments/assets/d25ff215-e804-4157-addb-9940c21b27e8" width=200>

If we filter out the parking the card stays.

You can take advantage of the zoom listener to fix the first problem and the launchedEffect on the filteredParkings for the second !
Hf

Armetiis commented:
> When we zoom with double click the card is not removed, and so we can move out and the card floats around without being on top of the parking 

Fixed! Now if the zoomLevel changes, the card gets disabled


> If we filter out the parking the card stays.

I changed this. Now the preview card is disabled if the parking gets filtered out from the selected parking, but it doesn't if the list gets updated but the parking is still in it :)

Thanks for the suggestions. Little details like this make it perfect!

sean-prz reviewed (approved):
Thank you, Perfect !

Pijukai reviewed (approved):
Thank you for fixing this bug ! The overall quality of the PR is really good and apart from the bug even the cc goes above 80% for the map. Great Job !
